src/main/java/xyz/nucleoid/plasmid/util/BucketFind.java

```diff
@@ -0,0 +1,44 @@
+package xyz.nucleoid.plasmid.util;
+
+import net.minecraft.block.BlockState;
+import net.minecraft.server.world.ServerWorld;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.util.math.Direction;
+
+import java.util.HashSet;
+import java.util.Set;
+import java.util.function.Predicate;
+
+public final class BucketFind {
+    /**
+     * Finds any directly connected blocks and puts them in a set.
+     *
+     * @param origin The position of the first block
+     * @param predicate The predicate for the blocks that can be accepted in the set
+     * @param limit The amount of maximum blocks to find
+     */
+    public static Set<BlockPos> find(ServerWorld world, BlockPos origin, Predicate<BlockState> predicate, int limit) {
```

Comment:
An additional `findByTag` method which allows passing a `Tag<Block>` rather than a `Predicate<BlockState>` would be helpful.